teps/0130-pipeline-level-service.md

```diff
@@ -0,0 +1,764 @@
+(stored on the hub, on the cluster, in git, etc, just like Tasks).
+
+However, the lifespan of a reusable sidecar wouldn't be tied to the lifespan of a PipelineRun,
+and managing it would expand the scope of responsibility of the PipelineRun controller.
+This alternative would only be scoped to sidecars defined inline in a Pipeline.
+
+### First-class Docker support
+
+We could build in first class support for Docker, for example, a user could specify that they'd like their Pipeline to have a Docker daemon provided.
+This solution is not proposed because it is too opinionated to align with our design principle on "flexibility".
+There are multiple ways of building images (e.g. kaniko, Docker, buildkit) and no reason that Tekton should prefer one over the others.
+Pipeline-level services can be used in much more flexible ways.
+
+### Docker daemon Custom Task
+
+We could create a Docker daemon custom task, for example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+spec:
+  workspaces:
+  - name: source
+  - name: docker-tls-certs
+  tasks:
+  - name: git-clone
+    ...
+  - name: startup-docker-daemon
+    taskRef:
+      apiVersion: tekton.dev/v1beta1
+      kind: docker
+    workspaces:
+    - name: docker-tls-certs
+  - name: docker-build
+    runAfter: ["startup-docker-daemon"]
+    params:
+    - name: docker-host
+      value: $(tasks.startup-docker-daemon.results.host)
+    ...
+    # Same spec as in the above example
+  - name: pytest
+    ...
+    # Same spec as in the above example
+  - name: docker-push
+    ...
+    # Same spec as in the above example
+```
+
+The Docker daemon custom task would be responsible for creating a service and deployment. Once the deployment is ready,
+the CustomRun would have "ConditionSucceeded" set to "true", and subsequent tasks would start.
+The custom task controller would also observe PipelineRuns, and delete any deployments and services associated with completed PipelineRuns.
+
+We could add support to the Catalog for custom tasks, and put the docker custom task into the Catalog.
+Work required to support custom tasks in the Catalog is tracked in https://github.com/tektoncd/catalog/issues/1147.
+
+Pros:
+- Compared to using regular Tasks, only one additional Task is needed (instead of setup and teardown)
+- Service account that can create deployments and services is only needed by the custom task controller instead of by each build Pipeline.
+The service account and rolebindings can be set up by the cluster operator when the custom task controller is installed, instead of app developers
+needing to understand why and how to use it in their Pipelines.
+- Do not need to copy/paste docker daemon container yaml into multiple Pipelines
+
+Cons:
+- Expands scope of the proposal to also include catalog support for custom tasks
+- Poor separation of concerns between PipelineRun controller and docker custom task controller, since the PipelineRun controller reacts to changes
```

Comment:
The pipelinerun controller already watches CustomRuns. I don't see why the CustomTask controller watching PipelineRun changes is a negative - Results watcher and other API clients do this too. If the custom task controller was updating the pipelinerun then I'd say it's an issue

Reply:
fair enough, I added custom task to one of the main options under consideration